### PR TITLE
enhancement(tls) expose peer TLS certificate information

### LINKED ISSUE #2931: Client certificate metadata option for sources
Hey all! Being able to propagate client certificate information as event metadata is a _really_ useful feature that I've used a lot with logstash in the past. From what I can tell, there isn't a capability to pass along that metadata with vector, at least with the socket and http input plugins that I could find.

It's a helpful feature for being able to collect metrics on specific users/applications while also providing some rudimentary ACLs for outputs when used in conjunction with the [translate](https://www.elastic.co/guide/en/logstash/current/plugins-filters-translate.html) filter, not to mention the added security benefit of using TLS.

Here's an example pipeline I've used before to map a beats input source to a given kafka topic based on the certificate principal they present(great for handling many distinct users over a single port):
```perl
input {
  beats {
    port => 7000
    ssl => true
    ssl_verify_mode => "force_peer"
    ssl_certificate_authorities => [
      "/etc/ssl/beats_server/ca.pem"
    ]
    ssl_certificate => "/etc/ssl/beats_server/certificate.pem"
    ssl_key => "/etc/ssl/beats_server/certificate-key_pkcs8.pem"
    ssl_peer_metadata => true
  }
}
filter {
  translate {
    field => "[@metadata][tls_peer][subject]"
    destination => "[@metadata][destination_topic]"
    dictionary_path => "/etc/logstash/cert_to_topic.yml"
    fallback => "catchall.topic"
    exact => true
  }
}

output {
  kafka {
    ...
    topic_id => "%{[@metadata][destination_topic]}"
  }
}
```

Where the dictionary path in the filter is a file That consists of a mapping of cert principal strings and topic names.

Not really sure how vector handles internal event metadata, but I was curious if you all would be open to supporting client certificate principal data to be extracted from incoming TLS connections(probably with the `socket` and `http` sources at first, but if there's interest maybe this feature could trickle out into some other sources, like the HEC input?) and maybe a translate-like transform(though I'm happy to open another issue for tracking the discussion on that.

What are your thoughts?

### PR DESCRIPTION
**Pull request by JustinKnueppel**
Closes #2931, this PR exposes the subject name of the peer TLS certificate of the incoming connection. Currently this PR exposes the certificate metadata in four sources - `socket`, `fluentd`, `syslog`, and `logstash`. Currently, there are a few unit tests to make sure the subject name is being formatted correctly for a given set of metadata, however we also have another branch that has testing on how this is exposed to the four listed sources. That testing is initially omitted in this PR as it changes more method signatures and has more refactoring that may make it more difficult to asses the core changes themselves. If there is any feedback or pointers for how to test this properly or if we should just add the tests we have and go from there please let us know!

cc @rwaweber

### CODE REVIEW

**Review by StephenWakely — CHANGES_REQUESTED** [review]
Thanks for doing this! The code is great, but I think that it should take a different general approach. 

It makes more sense that the certificate information should be added to the event metadata and not to the actual event itself. The metadata is a fairly new addition to Vector where information can be stored about the event, but does not leak into the final data that is output from the sink.

I'll add a separate post to highlight where I think the changes need to be made (sorry to be a pain!).

**Comment by StephenWakely** [comment]
We should add the field [here](https://github.com/JustinKnueppel/vector/blob/tls-cn/lib/vector-core/src/event/metadata.rs#L14-L30):

```
tls_certificate: Option<Arc<str>>
```

The `decoder` method can be modified to have a `CertificateMetadata` parameter which can be passed [here]( https://github.com/JustinKnueppel/vector/blob/tls-cn/src/sources/util/tcp/mod.rs#L284)

Then, eg. the [FluentDecoder](https://github.com/JustinKnueppel/vector/blob/tls-cn/src/sources/fluent/mod.rs#L101) can store this value in a field (it should probably call `to_string` here and store the string to avoid having to do this on every event). 

This can then be passed along until we reach [this point](https://github.com/JustinKnueppel/vector/blob/tls-cn/src/sources/fluent/mod.rs#L443) and instead of calling `LogEvent::default()` we can then call [`LogEvent::new_with_metadata`](https://github.com/JustinKnueppel/vector/blob/tls-cn/lib/vector-core/src/event/log_event.rs#L67) - the Metadata parameter being populated with our `CertificateMetadata` string.

To allow VRL to access this metadata field, we want to add the key [here](https://github.com/JustinKnueppel/vector/blob/tls-cn/lib/vector-vrl-functions/src/lib.rs#L8-L10) and to the functions in the target [here](https://github.com/JustinKnueppel/vector/blob/tls-cn/lib/vector-vrl-functions/src/lib.rs#L8-L10), [here](https://github.com/JustinKnueppel/vector/blob/tls-cn/lib/vector-core/src/event/vrl_target.rs#L285) and [here](https://github.com/JustinKnueppel/vector/blob/tls-cn/lib/vector-core/src/event/vrl_target.rs#L304).

That should cover it, if I've missed out any detail, or if you have any questions please give me a shout!

**Comment by jszwedko** [comment]
Hi @JustinKnueppel !

Thank you for this contribution! A number of users have asked for this so they will be happy to see it 😄 

I've been talking with @StephenWakely and we may actually want to keep the fields on the event itself rather than the metadata. I can follow up soon, but you could hold off on making that switch for now if you want (but can still respond to the other feedback here).

**Comment by JustinKnueppel** [comment]
> Hi @JustinKnueppel !
> 
> Thank you for this contribution! A number of users have asked for this so they will be happy to see it 😄
> 
> I've been talking with @StephenWakely and we may actually want to keep the fields on the event itself rather than the metadata. I can follow up soon, but you could hold off on making that switch for now if you want (but can still respond to the other feedback here).

Commits have been added for removing the unneeded helper function, and I have a little bit of work done to transition the change to the metadata that I'll stash in case that is the determination! Let me know what you all think is best.

**Comment by StephenWakely** [comment]
@JustinKnueppel Ok, sorry I jumped the gun a bit on the metadata front. It is a new feature and we haven't fully formalised how it should be used.

So at this point the field can be added to the actual event as you have already done. There's just a number of changes to make.

1. Adding the field should be configurable and it would be nice to be able to specify the name of the field that it is added to. It would be useful to add this option to the [TlsConfig struct](https://github.com/JustinKnueppel/vector/blob/tls-cn/src/tls/settings.rs#L35-L39) 

The issue here is that TlsConfig is used for sources as well as sinks, so you will need to create a `TlsSourceConfig` that wraps `TlsConfig`:

```rust
#[derive(Clone, Debug, Default, Deserialize, Serialize)]
pub struct TlsSourceConfig {
    pub peer_key: Option<String>,
    #[serde(flatten)]
    pub tls_config: TlsConfig,
}
```

Then the sources in question can use this instead of `Option<TlsConfig>`.

This would allow the config to look like:

```toml
[sources.fluent]
type = "fluent"
tls.peer_key = "certificate_metadata"
...
```

2. When adding the field it would be useful to add it to the event as a map. In future it could be useful to add more data here describing the connection.

Something like the following should do it in the `handle_events` function:

```rust
  if let Some(tls_peer_key) = self.tls_peer_key {
            if let Some(certificate_metadata) = certificate_metadata {
                let mut metadata = BTreeMap::new();
                metadata.insert("certificate", certificate_metadata.to_string());
                log.insert(tls_peer_key.clone(), metadata.into());
            }
}
```

You will need to add the `peer_key` field to the [`FluentSource`](https://github.com/JustinKnueppel/vector/blob/tls-cn/src/sources/fluent/mod.rs#L92) struct and pass the setting on in the [`build`](https://github.com/JustinKnueppel/vector/blob/tls-cn/src/sources/fluent/mod.rs#L59) method. It should be a similar pattern for the other sources. Thinking about it, it would be worth running the `to_string` here and storing the field as `Option<String>` in the `FluentSource`. This means we only have to run that once and not on every event. The String would still need cloning every event, so there would still be the overhead there, but it might save some small fractions of processing and it all counts!

That should cover it. Sorry for the earlier confusion. As always, any questions please give me a shout!

File: src/sources/fluent/mod.rs
Comment (jszwedko) [inline src/sources/fluent/mod.rs:114]: In addition to @StephenWakely's comment, I'd also move this `log.insert` call up to `handle_stream` in `src/sources/util/tcp/mod.rs` since we want it to apply to all `tcp` based sources. That should save some duplication here.

Comment (JustinKnueppel) [reply src/sources/fluent/mod.rs]: Just finished up an implementation of the above requested changes, let me know your thoughts!

File: src/sources/util/tcp/mod.rs
Comment (jszwedko) [inline src/sources/util/tcp/mod.rs:356]: Could we call this `subject`?
```suggestion
                                metadata.insert("subject".to_string(), certificate_metadata.to_string().into());
```

File: src/tls/incoming.rs
Comment (jszwedko) [inline src/tls/incoming.rs:393]: I might actually, instead of implementing `ToString` here, just have an implementation method called `subject` that returns the subject (as you have formatted here). I think that makes it clearer that that is what you are getting since the metadata could contain other info besides just the subject.

File: src/tls/incoming.rs
Comment (jszwedko) [inline src/tls/incoming.rs:399]: Typically you'd do `if let Some(cn) = &self.common_name` here instead of `match`.

File: src/tls/incoming.rs
Comment (jszwedko) [inline src/tls/incoming.rs:395]: I might do `with_capacity(6)` here instead of `new` to avoid reallocations.

**Review by jszwedko — COMMENTED** [review]
Great, thanks @JustinKnueppel !

I left some comments below, but one other thing it'd be nice to see here is an integration test that verifies it end-to-end. You could probably just augment this test to also verify the added `peer` fields:

https://github.com/vectordotdev/vector/blob/0a946cf7719de3a45b7e17723a638e8dab5a8309/src/sources/socket/mod.rs#L379-L414

You'll need to modify the `send_lines_tls` function to use a client certificate though. You can generate one and add it to `tests/data/`.

We also need to add this to the docs in `website/cue` but we can help with that.

File: src/tls/incoming.rs
Comment (bruceg) [reply src/tls/incoming.rs:399]: Since you're doing nothing in the `None` branch, using `if let` is simpler here too:
```suggestion
        if let Some(cn) = &self.common_name {
            components.push(format!("CN={}", cn)),
        }
```

Comment (bruceg) [inline src/tls/incoming.rs:399]: Another pattern here could be to write to a single `String` and trim the single trailing comma after:
```suggestion
        let mut result = String::new();
        if let Some(cn) = self.common_name.as_ref() {
            let _ = write!(&mut result, "CN={},", cn));
        }
        …
        result.pop();
        result
```

**Comment by JustinKnueppel** [comment]
Addressed the fixes requested above, and took a stab at the integration test additions. We've been trying to troubleshoot the issue here - https://discord.com/channels/742820443487993987/964208977942544395 as there seems to be some issue with getting the certificates to talk nicely. If there are any ideas for what we may be doing wrong with certificate creation please let us know!

File: src/sources/fluent/mod.rs
Comment (jszwedko) [reply src/sources/fluent/mod.rs:66]: I still think we could apply these `map`s.

**Review by jszwedko — APPROVED** [review]
Looks good, thanks for your work on this @JustinKnueppel !

If the integration tests are still failing, I can take a peek next week.

**Comment by JustinKnueppel** [comment]
@jszwedko Changed that to a map! And the integration test is indeed still failing and we'd love some help when you get some time! I think that is the last real part of this left.

**Comment by spencergilbert** [comment]
Sorry for the delays here @JustinKnueppel - I'll be taking a look at the integration test this week 👍

**Comment by spencergilbert** [comment]
Dropping myself a note that we still need to add docs for this as well

**Comment by jszwedko** [comment]
I think this is ready for re-review if you wouldn't mind taking a look @spencergilbert 

The biggest changes are:

* The test needed to configure the source to require the client certificate for the client to send it, and thus for the `peer_certificate` to be available.
* I regenerated all of the certificate assets as the current CA lacked the correct key usages. See the updated Makefile in `tests/data/Makefile` for generation of the new certificates.

**Comment by jszwedko** [comment]
And @StephenWakely too.

File: src/sources/socket/mod.rs
Comment (spencergilbert) [inline src/sources/socket/mod.rs:443]: Why are we dropping this test?

File: tests/data/Makefile
Comment (spencergilbert) [inline tests/data/Makefile:19]: RIP timber

**Review by spencergilbert — APPROVED** [review]

**Review by StephenWakely — DISMISSED** [review]
This is looking really good. There are just a number of `dbg` statements that need removing. Unfortunately there are a bunch of conflicts now, which I appreciate is frustrating.

**Comment by spencergilbert** [comment]
I can clean this up and get it ready to merge, most (all?) of those dbg's were mine and I can fixup the conflicts

**Comment by JustinKnueppel** [comment]
> I can clean this up and get it ready to merge, most (all?) of those dbg's were mine and I can fixup the conflicts

Saw this just as I pushed the branch! Debug statements are gone, but if you are willing to tackle the conflicts that would be fantastic!

File: website/cue/reference/components.cue
Comment (jszwedko) [inline website/cue/reference/components.cue:623]: I neglected this before, but should we add some notes about what will be added to the event when this key is set? Components have output descriptions too, like https://vector.dev/docs/reference/configuration/sources/socket/#output-logs , which seemingly it'd be nice to integrate this with.

Comment (spencergilbert) [reply website/cue/reference/components.cue:623]: Mhmm, your choice on if this is blocking - I'd like to move onto the ack work I have, but don't want the docs to rot

File: src/sources/vector/v1.rs
Comment (neuronull) [inline src/sources/vector/v1.rs:88]: This pattern appears to be found in most of the modified sources here (at least four times).
Would it be feasible and advantageous to extract this into a helper function say in `sources/tls/mod.rs` , which returns something like `crate::Result<(tls, tls_client_metadata_key)>` ?

Comment (spencergilbert) [reply src/sources/vector/v1.rs:88]: Doesn't sound like a bad idea, but I also don't love holding up this PR any more than necessary - it's been dragging for a while. I defer to @jszwedko

Comment (neuronull) [reply src/sources/vector/v1.rs:88]: Just a minor suggestion, no need to hold the PR just for this comment.

**Comment by spencergilbert** [comment]
does this need master merged in again - integration tests seem unhappy

**Comment by jszwedko** [comment]
@spencergilbert I think this is ready for another 👀 . I had to update the certs for some more of the integration tests. The Datadog integration tests fail because they don't have access to the `TEST_DATADOG_API_KEY` secret in CI due to this being a contributor PR.

**Review by spencergilbert — APPROVED** [review]
😅 this PR spiraled out into some many files, LGTM

**Comment by jszwedko** [comment]
Thanks for this contribution @JustinKnueppel !

**Pull request merged**
